### ISSUE TITLE
Exceeding character limit in Monster Hunter games causes weird overlapping.

### ISSUE THREAD

**Issue by tywald** [hrydgard/ppsspp#9821]
This bug is old but I finally know what causes it.

Basically, you can still add one more character when you hit the character limit and when you do, additional symbols are added and can go off screen.

That's how names can become like this:
![ules01213_00007](https://user-images.githubusercontent.com/10337060/27731892-3ad6a6cc-5d8f-11e7-8b0c-8c2fd06ea78c.png)
![ules01213_00008](https://user-images.githubusercontent.com/10337060/27731897-40cf9066-5d8f-11e7-85d6-6a1b4155741e.png)
![ules01213_00009](https://user-images.githubusercontent.com/10337060/27731896-40bbf5e2-5d8f-11e7-8952-1e4d1bab7db6.png)

It looks like it would stop at 8 but as you can see in the next screenshot you can have 9:
![ules01213_00014](https://user-images.githubusercontent.com/10337060/27771458-fa9cc566-5f4e-11e7-9c7f-931626b63c26.png)
Max characters should be 8, but you can have 9(which shouldn't be allowed):
![ules01213_00013](https://user-images.githubusercontent.com/10337060/27770909-5cce7d40-5f47-11e7-9e08-77dacb7f689b.png)

Same thing can happen when you add a comment to a guild card:
Just as you hit character limit(working as intended):
![ules01213_00010](https://user-images.githubusercontent.com/10337060/27731904-4a7fb60e-5d8f-11e7-82a7-2c2461a7955f.png)

Exceeding:
![ules01213_00011](https://user-images.githubusercontent.com/10337060/27731907-4f4af2b6-5d8f-11e7-9ef1-34417b208439.png)
Was overlapped from:
![ules01213_00012](https://user-images.githubusercontent.com/10337060/27731921-5bf93aa4-5d8f-11e7-8d08-f76b27639cd6.png)

![uljm05800_00194](https://user-images.githubusercontent.com/10337060/27771129-f2bd673e-5f48-11e7-805e-d8a3f025a659.png)
![uljm05800_00195](https://user-images.githubusercontent.com/10337060/27771131-f4d89f98-5f48-11e7-96e4-ec603f125b06.png)

**Comment by hrydgard** [comment]
Okay, so text input character limit is either one too long or doesn't zero-terminate properly. Shouldn't be too difficult to fix, adding to 1.5 milestone.

**Comment by unknownbrackets** [comment]
Can you test this on a real PSP - does it let you type 17 characters?

-[Unknown]

**Comment by tywald** [comment]
I had a couple of friends who could assist me to test on a PSP.

For naming a character in Monster Hunter Freedom Unite, it didn't let them go past 8 characters.

![jpeg_20170702_120611](https://user-images.githubusercontent.com/10337060/27771643-55119c02-5f53-11e7-9d93-58c139eb2039.jpg)
![jpeg_20170702_120346](https://user-images.githubusercontent.com/10337060/27771649-69e2ea3c-5f53-11e7-863a-37ab3917aa86.jpg)

For guild card comments, it didn't let them go past 16 characters.
![img_20170702_124637](https://user-images.githubusercontent.com/10337060/27772060-92ef9454-5f5b-11e7-917b-b5cfdb1d1c5b.jpg)
![img_20170702_130419](https://user-images.githubusercontent.com/10337060/27772061-9780c4a2-5f5b-11e7-8b26-f38e11fc11b9.jpg)

That's the maximum which was allowed.

**Comment by unknownbrackets** [comment]
We currently have two fields:

```c++
	// Length, in unsigned shorts, including the terminator.
	u32_le outtextlength;
	// Number of characters to allow, not including terminator (if less than outtextlength - 1.)
	u32_le outtextlimit;
```

And also, #3920 adjusted these to make password entry work correctly in Valkyria Chronicles.

We should be writing the null terminator:

```c++
	size_t end = oskParams->fields[0].outtextlength;
	// Only write the bytes of the output and the null terminator, don't write the rest.
	if (end > inputChars.size())
		end = inputChars.size() + 1;
```

I don't have this game, but I wonder what parameters it's passing for those two length fields.

-[Unknown]

**Comment by tywald** [comment]
Naming a character in Monster Hunter Portable 3rd and in the HD version seems to be fine, it uses a 12 character space and it doesn't let you go over those 12 characters. Still the same issue with the 16 character space for the guild card comments however as seen in the last two screenshots in the original post.

Naming a character in Monster Hunter Freedom which has a 8 character limit the same as in Freedom Unite also exceeds it.

![ules00318_00400](https://user-images.githubusercontent.com/10337060/27779369-d2ac8a7c-5fc2-11e7-8e22-fde646471094.png)
![ules00318_00399](https://user-images.githubusercontent.com/10337060/27779370-d2b053a0-5fc2-11e7-80a8-ee75636abb79.png)

So basically:
8 character space - problem.
12 character space - working fine.
16 character space - problem.

**Comment by vnctdj** [comment]
Cool, thanks @unknownbrackets!
